docs/semgrep-code/java.md

```diff
@@ -0,0 +1,80 @@
+
+This demo rule detects tainted data in `sink`. 
+
+* This example has two true positives: **line 22** and **line 28**.
+* Semgrep Pro is able to detect that **line 24 and 30 are false positives**. Semgrep OSS can't catch that distinction.
+* Line 24 is a false positive because the data in the sink is an element of an integer List.
+* Line 30 is a false positive because the data in the sink is a set of booleans.
+* The Semgrep rule uses the fields `taint_assume_safe_booleans` and `taint_assume_safe_numbers` to tell the engine that these types are safe and not tainted. 
+
+## Semgrep understands the Java standard library and APIs
+
+Java provides a wide array of standard classes and methods across its various libraries. These facilitate programming by offering ready-to-use methods for common tasks. Many of these take string inputs, and return integer or boolean values. Thus, these statements returning integer or boolean values are not considered tainted. Semgrep is able to make that distinction, preventing this type of false positive. 
+
+### Example: `sqli-demo-bool_doesnt_taint`
+
+<iframe title="tk" src="https://semgrep.dev/embed/editor?snippet=Kx1AY" width="100%" height="432px" frameBorder="0"></iframe>
+
+**Figure**. `sqli-demo-bool_doesnt_taint`. To view the entire sample code and rule, click **Open in Playground**.
+
+This demo rule detects SQL injection through a `UserInputGenerator` class. This class's unsanitized user input is passed to `SQLQueryRunner`.
+
+* This example has two true positives: **line 11** and **line 20**.
+* Semgrep Pro is able to detect that **line 14 and 17 are false positives**. Semgrep OSS can't catch that distinction.
+* Lines 14 and 17 are false positives because `input.endsWith("something")` and `input.indexOf('u')` return a boolean and integer respectively. Semgrep Pro is able to understand `endsWith` and `indexOf` Java methods.
+* The Semgrep rule uses the fields `taint_assume_safe_booleans` and `taint_assume_safe_numbers` to tell the engine that these types are safe and not tainted. 
+
+## Target code in a parent class and its subclasses
+
+Semgrep supports class inheritance in Java. You can use Semgrep to search across all subclasses. This specificity means that rules can better target your codebase, increasing true positive rates. This is achieved through the `metavariable-type` field, which can accept the name of any user-defined class. 
+
+The `metavariable-type` field is available in Semgrep OSS. However, classes in Java are frequently defined across files (interfile), which is beyond the scope of Semgrep OSS's analysis. Use Semgrep Pro to perform cross-file analysis to ensure that Semgrep can detect all class and subclass definitions.
+
+### Example: `detect-pattern-in-subclass`
+
+<iframe title="tk" src="https://semgrep.dev/embed/editor?snippet=nJjjG" width="100%" height="432px" frameBorder="0"></iframe>
```

Comment:
I think in Java you can just write `(Foo $CLASS).x

Reply:
Yes but then it matches because it follows the Java syntax, rather than using the `type` key. I wanted to illustrate the `type` key in this case.